core/src/display_object/edit_text.rs

```diff
@@ -2312,3 +2328,210 @@ impl TextSelection {
+                    current_intervals.push((c, c));
+                    Some(c)
+                }
+                EditTextRestrictToken::Caret => {
+                    if now_allowing {
+                        if current_intervals.is_empty() && allowed.is_empty() {
+                            // If restrict starts with ^, we are assuming that
+                            // all characters are allowed and disallowing from that.
+                            allowed.append(&mut vec![Self::INTERVAL_ALL]);
+                        } else {
+                            allowed.append(&mut current_intervals);
+                        }
+                    } else {
+                        disallowed.append(&mut current_intervals);
+                    }
+
+                    // Caret according to the documentation indicates
+                    // that we are now disallowing characters.
+                    // In reality it just switches allowing/disallowing.
+                    now_allowing = !now_allowing;
+                    None
+                }
+                EditTextRestrictToken::Range => {
+                    let range_start = if let Some(last_char) = last_char {
+                        current_intervals.pop();
+                        last_char
+                    } else {
+                        // When the range is truncated from the left side (-z),
+                        // it is equivalent to \0-z.
+                        '\0'
+                    };
+                    let range_end;
+                    if let Some(EditTextRestrictToken::Char(c)) = tokens.front() {
+                        range_end = *c;
+                        tokens.pop_front();
+                    } else {
+                        // When the range is truncated from the right side (a-),
+                        // it is equivalent to the first character (a).
+                        range_end = range_start;
+                    }
+                    // If the range a-z is inverted (z-a), it is equivalent to
+                    // the first character only (z).
+                    current_intervals.push((range_start, range_end.max(range_start)));
+                    None
+                }
+            }
+        }
+
+        if now_allowing {
+            allowed.append(&mut current_intervals);
+        } else {
+            disallowed.append(&mut current_intervals);
+        }
+    }
+
+    pub fn value(&self) -> Option<&WStr> {
+        self.value.as_deref()
+    }
+
+    pub fn is_allowed(&self, character: char) -> bool {
+        self.intervals_contain(character, &self.allowed)
+            && !self.intervals_contain(character, &self.disallowed)
+    }
+
+    fn intervals_contain(&self, character: char, intervals: &Vec<(char, char)>) -> bool {
+        for interval in intervals {
+            if self.interval_contains(character, interval) {
+                return true;
+            }
+        }
+        false
```

Comment:
I guess this could have been a call to `any`, but this isn't any less readable, and no need to delay further just for this.